R/hdf5.R

```diff
@@ -1,110 +1,207 @@
+                      md_fl == md
+                  })
+    all(unlist(res))
+}
+
+#' @title The `Hdf5MSnExp` Class for MS Data And Meta-Data
+#'
+#' @description
+#'
+#' The `Hdf5MSnExp` class encapsulates data and meta-data for mass
+#' spectrometry experiments like the `MSnExp` and `OnDiskMSnExp`
+#' classes. `Hdf5MSnExp` implements, like `OnDiskMSnExp`, an
+#' *on disk* model, where the raw data (M/Z and intensities) are
+#' stored on disk (rather than in memory, like `MSnExp` data) using,
+#' as the name implies, the hdf5 file format. See [`MSnExp`] and
+#' [`OnDiskMSnExp`] for a more general description of the data,
+#' slots and operations.
+#'
+#' Object of the class are currently created with the
+#' `readHdf5DiskMSData` function. Later, this backend will be
+#' included in the main `readMSData` function.
+#'
+#' @slot hdf5file `character` containing the hdf5 filenames, one file
+#'     per input mzML file/sample.
+#'
+#' @rdname Hdf5MSnExp-class
+#'
+#' @aliases Hdf5MSnExp
+#'
+#' @seealso See [`MSnExp`] and [`OnDiskMSnExp`] classes.
+#'
+#' @author Laurent Gatto, Johannes Rainer
+#'
+#' @md
+#'
+#' @examples
+#' f <- msdata::proteomics(pattern = "MS3TMT11", full.names = TRUE)
+#' x <- readHdf5DiskMSData(f, hdf5path = tempdir())
+#' x
+#'
+#' ## automatically generated hdf5 file
+#' hdf5FileName(x)
+#'
+#' x[[1]]
+#' x[[2]]
+#' x[[10]]
+#' filterMsLevel(x, 3L)[[1]]
+#'
+#' ## Perform data manipulations, remove all peaks with an intensity
+#' ## smaller than 1000
+#' x_clean <- clean(removePeaks(x, t = 1000), all = TRUE)
+#'
+#' x_clean[[12]]
+#'
+#' ## Data manipulations are, as for OnDiskMSnExp objects, performed
+#' ## on-the-fly. The data in the hdf5 files was not changed by the
+#' ## clean and removePeaks calls above. Object x is thus still valid
+#' ## and we can still use it.
+#' MSnbase:::validHdf5MSnExp(x)
+#' x[[12]]
+#'
+#' ## updateHdf5Data will make all data manipulations persistent, it will
+#' ## overwrite the data in the hdf5 files associated with x_clean and
+#' ## accessing the data afterwards will no longer require to apply
+#' ## data manipulations on-the-fly.
+#' updateHdf5Data(x_clean)
+#'
+#' x_clean[[12]]
+#'
+#' ## Note that the original object x is however still associated with the
+#' ## same hdf5 files. This object will now no longer be valid, as the data
+#' ## has changed for it.
+#' MSnbase:::validHdf5MSnExp(x)
+#'
+#' ## Accessing the data such as with x[[12]] will thus result in an error.
+#' ## The data can however be *restored* with the `convertToHdf5MSnExp`
+#' ## function that will re-read all data into hdf5 files. Note that a
+#' ## different `hdf5path` has to be used.
+#' x <- convertToHdf5MSnExp(x, hdf5path = paste0(tempdir(), "/2/"))
+#'
+#' MSnbase:::validHdf5MSnExp(x)
+#' x[[12]]
+#'
+#' ## clean up session
+#' file.remove(hdf5FileName(x))
+.Hdf5MSnExp <- setClass("Hdf5MSnExp",
+                        slots = c(hdf5file = "character",
+                                  md5sum = "character"),
```

Comment:
slot `md5sum` stores the md5 sums of the spectra within a hdf5 file. The same number is stored into the file upon generation of the hdf5 file. Also, `updateHdf5MSnExp` updates both the md5 sum in the file and the one in this slot. If they are not the same, the hdf5 file has been changed.